// @minka/ledger-sdk/src/lib/domain/domain-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DomainClient = exports.DomainsResponse = exports.DomainResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const DOMAINS_PATH = '/domains';
/**
 * Single domain API response.
 */
class DomainResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    domain;
    constructor(response) {
        super(response);
        this.domain = response?.data?.data;
    }
}
exports.DomainResponse = DomainResponse;
/**
 * Domains list API response.
 */
class DomainsResponse extends list_response_1.BaseListResponse {
    /**
     * A flattened list of records, extracted from the `data` property of
     * each record returned.
     */
    domains;
    constructor(response) {
        super(response);
        this.domains = response?.data?.data?.map((record) => record.data);
    }
}
exports.DomainsResponse = DomainsResponse;
/**
 * Allows querying and managing domains.
 *
 * @see {LedgerDomain}
 */
class DomainClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, DOMAINS_PATH);
        this.access = new access_client_1.AccessClient(options, DOMAINS_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     *
     * @see {LedgerDomain}
     * @see {CreateRecordBuilder}
     * @param options.record initial record payload (optional)
     * @param options.handle specify handle to be updated, if any (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(DOMAINS_PATH, DomainResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     *
     * @see {LedgerDomain}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${DOMAINS_PATH}/${encodeURIComponent(record.data.handle)}`, DomainResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${DOMAINS_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, DomainResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets a ledger domain by identifier (luid or handle).
     *
     * @see {LedgerDomain}
     * @throws {Error} if the domain doesn't exist or the
     *   current user doesn't have access to this domain
     * @param id a unique domain handle
     * @returns a response with the requested domain
     */
    async read(id, options) {
        return this.getRecord(`${DOMAINS_PATH}/${encodeURIComponent(id)}`, DomainResponse, options);
    }
    /**
     * Queries domains from a ledger.
     *
     * @see {LedgerDomain}
     * @param params query parameters
     * @returns domains list response with all found domains
     */
    async list(params) {
        return this.getRecordsList(DOMAINS_PATH, DomainsResponse, params);
    }
}
exports.DomainClient = DomainClient;
