// @minka/ledger-sdk/src/lib/common/clients/access-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessClient = exports.AccessCheckResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const check_access_builder_1 = require("../services/check-access.builder");
const list_response_1 = require("../types/list-response");
const base_client_1 = require("./base-client");
const ACCESS_PATH = 'access';
/**
 * Access check API response.
 */
class AccessCheckResponse extends list_response_1.BaseListResponse {
    rules;
    constructor(response) {
        super(response);
        this.rules = response?.data?.data?.map((record) => record.data);
    }
}
exports.AccessCheckResponse = AccessCheckResponse;
/**
 * Allows querying access.
 *
 * @see {AccessRule}
 */
class AccessClient extends base_client_1.BaseClient {
    parentPath;
    constructor(options, parentPath) {
        super(options);
        this.parentPath = parentPath;
    }
    buildBasePath() {
        const address = this.assertParentId();
        return `${this.parentPath}/${encodeURIComponent(address)}/${ACCESS_PATH}`;
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     *
     * @see {LedgerAccessRule}
     * @see {AccessCheckBuilder}
     * @param record initial record payload (optional)
     * @returns AccessCheckBuilder instance
     */
    check(record) {
        const builder = new check_access_builder_1.AccessCheckBuilder(async (record, reqOptions) => {
            return this.checkAccess(this.buildBasePath() + '/!check', AccessCheckResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
}
exports.AccessClient = AccessClient;
