// @minka/ledger-sdk/src/lib/metric/metric-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricClient = exports.MetricTimeseriesListResponse = exports.MetricsResponse = exports.MetricResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const base_client_1 = require("../common/clients/base-client");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const METRICS_PATH = '/metrics';
/**
 * Single metric API response.
 */
class MetricResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    metric;
    constructor(response) {
        super(response);
        this.metric = response?.data?.data;
    }
}
exports.MetricResponse = MetricResponse;
/**
 * Metrics list API response.
 */
class MetricsResponse extends list_response_1.BaseListResponse {
    /**
     * A flattened list of records, extracted from the `data` property of
     * each record returned.
     */
    metrics;
    constructor(response) {
        super(response);
        this.metrics = response?.data?.data?.map((record) => record.data);
    }
}
exports.MetricsResponse = MetricsResponse;
/**
 * Metrics list API response.
 */
class MetricTimeseriesListResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * A flattened list of records, extracted from the `data` property of
     * each record returned.
     */
    timeseries;
    constructor(response) {
        super(response);
        this.timeseries = response?.data?.data;
    }
}
exports.MetricTimeseriesListResponse = MetricTimeseriesListResponse;
/**
 * Allows querying and managing metrics.
 * @see {LedgerMetric}
 */
class MetricClient extends base_client_1.BaseClient {
    constructor(options) {
        super(options);
    }
    /**
     * Gets a ledger metric by identifier (luid or handle).
     * @see {LedgerMetric}
     * @param options
     * @throws {Error} if the metric doesn't exist or the
     *   current user doesn't have access to this metric
     * @param id a unique metric handle
     * @returns a single metric response with the requested metric
     */
    async read(id, options) {
        return this.getRecord(`${METRICS_PATH}/${encodeURIComponent(id)}`, MetricResponse, options);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     * @see {LedgerMetric}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${METRICS_PATH}/${encodeURIComponent(record.data.handle)}`, MetricResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${METRICS_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, MetricResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Queries metrics from a ledger.
     * @see {LedgerMetric}
     * @param id
     * @param options
     * @param params query parameters
     * @returns metrics list response with all found metrics
     */
    async listTimeseries(id, options) {
        return this.getRecord(`${METRICS_PATH}/${encodeURIComponent(id)}/timeseries`, MetricTimeseriesListResponse, options);
    }
}
exports.MetricClient = MetricClient;
