// @minka/ledger-sdk/crypto/types/src/lib/records/ledger-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestAction = exports.RequestRecord = exports.RequestSchema = exports.RequestParticipant = void 0;
/**
 * RequestParticipant is the prefix for the RequestParticipantHandle field,
 * except for `unknown` which is used alone when the participant is not known.
 */
var RequestParticipant;
(function (RequestParticipant) {
    RequestParticipant["Unknown"] = "unknown";
    RequestParticipant["Ledger"] = "ledger";
    RequestParticipant["Signer"] = "signer";
    RequestParticipant["Bridge"] = "bridge";
    RequestParticipant["Effect"] = "effect";
})(RequestParticipant || (exports.RequestParticipant = RequestParticipant = {}));
/**
 * RequestSchema represents the type of request that is logged.
 * At the moment we only support REST requests, but additional ones are
 * possible like Kafka.
 */
var RequestSchema;
(function (RequestSchema) {
    RequestSchema["Rest"] = "rest";
})(RequestSchema || (exports.RequestSchema = RequestSchema = {}));
/**
 *  RequestRecord is the prefix for the LedgerRequest
 *  record field which has the format prefix:id. In case there is no
 *  specific id like in the case of querying records, just the prefix is used.
 */
var RequestRecord;
(function (RequestRecord) {
    RequestRecord["AdviceAliasResolve"] = "advice-alias-resolve";
    RequestRecord["Anchor"] = "anchor";
    RequestRecord["AnchorChange"] = "anchor-change";
    RequestRecord["AnchorProof"] = "anchor-proof";
    RequestRecord["AnchorAccess"] = "anchor-access";
    RequestRecord["Bridge"] = "bridge";
    RequestRecord["BridgeChange"] = "bridge-change";
    RequestRecord["BridgeProof"] = "bridge-proof";
    RequestRecord["BridgeAccess"] = "bridge-access";
    RequestRecord["Circle"] = "circle";
    RequestRecord["CircleChange"] = "circle-change";
    RequestRecord["CircleProof"] = "circle-proof";
    RequestRecord["CircleAccess"] = "circle-access";
    RequestRecord["CircleSigner"] = "circle-signer";
    RequestRecord["Domain"] = "domain";
    RequestRecord["DomainChange"] = "domain-change";
    RequestRecord["DomainProof"] = "domain-proof";
    RequestRecord["DomainAccess"] = "domain-access";
    RequestRecord["Effect"] = "effect";
    RequestRecord["EffectChange"] = "effect-change";
    RequestRecord["EffectProof"] = "effect-proof";
    RequestRecord["EffectAccess"] = "effect-access";
    RequestRecord["Entry"] = "entry";
    RequestRecord["Event"] = "event";
    RequestRecord["Intent"] = "intent";
    RequestRecord["IntentChange"] = "intent-change";
    RequestRecord["IntentProof"] = "intent-proof";
    RequestRecord["IntentAccess"] = "intent-access";
    RequestRecord["Ledger"] = "ledger";
    RequestRecord["LedgerChange"] = "ledger-change";
    RequestRecord["LedgerProof"] = "ledger-proof";
    RequestRecord["LedgerAccess"] = "ledger-access";
    RequestRecord["Metric"] = "metric";
    RequestRecord["Policy"] = "policy";
    RequestRecord["PolicyChange"] = "policy-change";
    RequestRecord["PolicyProof"] = "policy-proof";
    RequestRecord["PolicyAccess"] = "policy-access";
    RequestRecord["Report"] = "report";
    RequestRecord["ReportChange"] = "report-change";
    RequestRecord["ReportProof"] = "report-proof";
    RequestRecord["ReportAccess"] = "report-access";
    RequestRecord["Request"] = "request";
    RequestRecord["Schema"] = "schema";
    RequestRecord["SchemaChange"] = "schema-change";
    RequestRecord["SchemaProof"] = "schema-proof";
    RequestRecord["SchemaAccess"] = "schema-access";
    RequestRecord["Server"] = "server";
    RequestRecord["ServerAccess"] = "server-access";
    RequestRecord["Signer"] = "signer";
    RequestRecord["SignerChange"] = "signer-change";
    RequestRecord["SignerProof"] = "signer-proof";
    RequestRecord["SignerAccess"] = "signer-access";
    RequestRecord["SignerFactor"] = "signer-factor";
    RequestRecord["SignerFactorSecret"] = "signer-factor-secret";
    RequestRecord["SignerFactorChange"] = "signer-factor-change";
    RequestRecord["SignerFactorProof"] = "signer-factor-proof";
    RequestRecord["SignerFactorAccess"] = "signer-factor-access";
    RequestRecord["Symbol"] = "symbol";
    RequestRecord["SymbolChange"] = "symbol-change";
    RequestRecord["SymbolProof"] = "symbol-proof";
    RequestRecord["SymbolAccess"] = "symbol-access";
    RequestRecord["Wallet"] = "wallet";
    RequestRecord["WalletChange"] = "wallet-change";
    RequestRecord["WalletProof"] = "wallet-proof";
    RequestRecord["WalletAccess"] = "wallet-access";
})(RequestRecord || (exports.RequestRecord = RequestRecord = {}));
/**
 * RequestAction represents what was done to/with the record.
 * For incoming REST requests Create corresponds to POST, Read to GET,
 * Update to PUT and Query to GET on the collection. Sign and ReadBalance
 * are actions that don't directly correspond to REST methods on the
 * record. The combination of RequestRecord and RequestAction
 * uniquely identify Ledger endpoints. Send is used when dispatching webhook
 * Events while Prepare, Commit and Abort are used for two-phase commit
 * requests.
 */
var RequestAction;
(function (RequestAction) {
    RequestAction["Abort"] = "abort";
    RequestAction["Activate"] = "activate";
    RequestAction["RetryEvent"] = "retry-event";
    RequestAction["QueryEvent"] = "query-event";
    RequestAction["AssignSigner"] = "assign-signer";
    RequestAction["Commit"] = "commit";
    RequestAction["Create"] = "create";
    RequestAction["Drop"] = "drop";
    RequestAction["Lookup"] = "lookup";
    RequestAction["Check"] = "check";
    RequestAction["Prepare"] = "prepare";
    RequestAction["Query"] = "query";
    RequestAction["QueryCircle"] = "query-circle";
    RequestAction["QuerySigner"] = "query-signer";
    RequestAction["Read"] = "read";
    RequestAction["ReadBalance"] = "read-balance";
    RequestAction["ReadLimit"] = "read-limit";
    RequestAction["ReadSigner"] = "read-signer";
    RequestAction["RemoveSigner"] = "remove-signer";
    RequestAction["QueryTimeseries"] = "query-timeseries";
    RequestAction["QueryValue"] = "query-value";
    RequestAction["Send"] = "send";
    RequestAction["Update"] = "update";
    RequestAction["Reveal"] = "reveal";
})(RequestAction || (exports.RequestAction = RequestAction = {}));
