// @minka/ledger-sdk/crypto/types/src/lib/records/ledger-policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuorumProofSelectionValue = exports.StatusPolicyConfigKey = exports.Weekday = exports.SchedulePolicyAction = exports.ProcessingAspectStrategy = exports.ProcessingAspectAction = void 0;
var ProcessingAspectAction;
(function (ProcessingAspectAction) {
    ProcessingAspectAction["Read"] = "read";
    ProcessingAspectAction["Query"] = "query";
    ProcessingAspectAction["Create"] = "create";
    ProcessingAspectAction["Update"] = "update";
    ProcessingAspectAction["Drop"] = "drop";
    ProcessingAspectAction["Sign"] = "sign";
})(ProcessingAspectAction || (exports.ProcessingAspectAction = ProcessingAspectAction = {}));
var ProcessingAspectStrategy;
(function (ProcessingAspectStrategy) {
    ProcessingAspectStrategy["Proxy"] = "proxy";
    ProcessingAspectStrategy["Fallback"] = "fallback";
    ProcessingAspectStrategy["Validate"] = "validate";
    ProcessingAspectStrategy["Synchronize"] = "synchronize";
})(ProcessingAspectStrategy || (exports.ProcessingAspectStrategy = ProcessingAspectStrategy = {}));
var SchedulePolicyAction;
(function (SchedulePolicyAction) {
    SchedulePolicyAction["Create"] = "create";
})(SchedulePolicyAction || (exports.SchedulePolicyAction = SchedulePolicyAction = {}));
var Weekday;
(function (Weekday) {
    Weekday["Mon"] = "mon";
    Weekday["Tue"] = "tue";
    Weekday["Wed"] = "wed";
    Weekday["Thu"] = "thu";
    Weekday["Fri"] = "fri";
    Weekday["Sat"] = "sat";
    Weekday["Sun"] = "sun";
})(Weekday || (exports.Weekday = Weekday = {}));
var StatusPolicyConfigKey;
(function (StatusPolicyConfigKey) {
    StatusPolicyConfigKey["QuorumProofSelection"] = "quorum.proofSelection";
})(StatusPolicyConfigKey || (exports.StatusPolicyConfigKey = StatusPolicyConfigKey = {}));
var QuorumProofSelectionValue;
(function (QuorumProofSelectionValue) {
    /**
     * Proofs are selected from the latest chain with target status (custom.status)
     */
    QuorumProofSelectionValue["LatestChain"] = "latest-chain";
    /**
     * Proofs are select from the entire set of proofs of latest version of record
     */
    QuorumProofSelectionValue["EntireSet"] = "entire-set";
})(QuorumProofSelectionValue || (exports.QuorumProofSelectionValue = QuorumProofSelectionValue = {}));
