// @minka/ledger-sdk/src/lib/server/server-access-client.js
"use strict";
/**
 * @module Clients / Server Access
 * @category Entities
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerAccessClient = void 0;
const access_client_1 = require("../common/clients/access-client");
const parse_access_builder_1 = require("../common/services/parse-access.builder");
const SERVER_ACCESS_PATH = 'access';
/**
 * Allows checking server access.
 *
 */
class AccessDelegateClient extends access_client_1.AccessClient {
    constructor(options) {
        super(options, '');
    }
    buildBasePath() {
        return SERVER_ACCESS_PATH;
    }
}
/**
 * Allows querying access at server level.
 */
class ServerAccessClient {
    accessClient;
    parseAccessBuilder;
    constructor(options) {
        this.accessClient = new AccessDelegateClient(options);
    }
    /**
     * Checks access at server level
     *
     * @param record access check
     * @returns builder instance
     */
    check(record) {
        return this.accessClient.check(record);
    }
    /**
     * Set config params to build authentication Jwt
     *
     * @param authParams jwt configuration
     */
    setAuthParams(authParams) {
        this.accessClient.setAuthParams(authParams);
    }
    /**
     * Sets header
     *
     * @param key
     * @param value
     */
    setHeader(key, value) {
        this.accessClient.setHeader(key, value);
    }
    /**
     * Initializes parse access builder with source access rules.
     * @param accessRules source access rules
     * @returns
     */
    with(accessRules) {
        return new parse_access_builder_1.ParseAccessBuilder(accessRules);
    }
}
exports.ServerAccessClient = ServerAccessClient;
