// @minka/ledger-sdk/crypto/types/src/lib/records/ledger/ledger-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForwardStrategy = exports.LedgerAccessStrategy = exports.LedgerConfigKey = void 0;
var LedgerConfigKey;
(function (LedgerConfigKey) {
    // Access
    LedgerConfigKey["AccessStrategy"] = "access.strategy";
    // Intent
    LedgerConfigKey["IntentExpiryThresholdMinutes"] = "intent.expiryThresholdMinutes";
    LedgerConfigKey["IntentExpiryThresholdMinutesDeprecated"] = "intentExpiryThresholdMinutes";
    LedgerConfigKey["IntentCreditorInitiatedTransfersEnabled"] = "intent.creditorInitiatedTransfersEnabled";
    // Domain
    LedgerConfigKey["DomainResolutionFromHandleEnabled"] = "domain.resolutionFromHandleEnabled";
    // Anchor
    LedgerConfigKey["AnchorWalletRequired"] = "anchor.walletRequired";
    // Policy
    LedgerConfigKey["PolicyScheduleProcessorEnabled"] = "policy.scheduleProcessorEnabled";
    // Traps
    LedgerConfigKey["TrapsEnabled"] = "traps.enabled";
    // Request
    LedgerConfigKey["RequestJournalingEnabled"] = "request.journalingEnabled";
    // Limits
    LedgerConfigKey["AggregatedLimitsEnabled"] = "limits.aggregated.enabled";
    // Authentication
    LedgerConfigKey["AuthenticationTwoFactorEnabled"] = "authentication.twoFactor.enabled";
    // Forwarding
    LedgerConfigKey["ForwardAnchorCreate"] = "forward.anchor.create";
    LedgerConfigKey["ForwardAnchorUpdate"] = "forward.anchor.update";
    LedgerConfigKey["ForwardAnchorDrop"] = "forward.anchor.drop";
    LedgerConfigKey["ForwardAnchorGet"] = "forward.anchor.get";
    LedgerConfigKey["ForwardAnchorSign"] = "forward.anchor.sign";
    LedgerConfigKey["ForwardAnchorFindAll"] = "forward.anchor.findAll";
    LedgerConfigKey["ForwardAnchorStrategy"] = "forward.anchor.strategy";
    LedgerConfigKey["ForwardAnchorCreateStrategy"] = "forward.anchor.create.strategy";
    LedgerConfigKey["ForwardAnchorUpdateStrategy"] = "forward.anchor.update.strategy";
    LedgerConfigKey["ForwardAnchorDropStrategy"] = "forward.anchor.drop.strategy";
    LedgerConfigKey["ForwardAnchorGetStrategy"] = "forward.anchor.get.strategy";
    LedgerConfigKey["ForwardAnchorSignStrategy"] = "forward.anchor.sign.strategy";
    LedgerConfigKey["ForwardAnchorFindAllStrategy"] = "forward.anchor.findAll.strategy";
})(LedgerConfigKey || (exports.LedgerConfigKey = LedgerConfigKey = {}));
var LedgerAccessStrategy;
(function (LedgerAccessStrategy) {
    LedgerAccessStrategy["RecordBased"] = "record-based";
    LedgerAccessStrategy["PolicyBased"] = "policy-based";
})(LedgerAccessStrategy || (exports.LedgerAccessStrategy = LedgerAccessStrategy = {}));
var ForwardStrategy;
(function (ForwardStrategy) {
    ForwardStrategy["None"] = "none";
    ForwardStrategy["Proxy"] = "proxy";
    ForwardStrategy["Fallback"] = "fallback";
    ForwardStrategy["Validate"] = "validate";
    ForwardStrategy["Synchronize"] = "synchronize";
})(ForwardStrategy || (exports.ForwardStrategy = ForwardStrategy = {}));
