// @minka/ledger-sdk/crypto/src/lib/node/signing.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.signHash = signHash;
exports.verifySignature = verifySignature;
const crypto = __importStar(require("crypto"));
const asserts_1 = require("../asserts");
const format_1 = require("./format");
async function signHash(hash, keyPair) {
    (0, asserts_1.assertKeyFormat)(keyPair.format);
    if (!crypto.sign) {
        throw new Error('Unable to sign hash, `crypto.sign` is not defined');
    }
    const hashBuffer = Buffer.from(hash, 'hex');
    const key = (0, format_1.importPrivateKey)(keyPair.secret);
    // The first argument must be undefined for ed25519, it defines a digest algorithm,
    // but ed25519 makes a sha512 digest as part of the algorithm. Signing and verification
    // with ed25119 doesn't work correctly if that argument is provided. Also, most crypto
    // examples use createSign(algorithm) to first create a Sign class and use that for
    // signing, but that way also doesn't work with ed25519.
    // see: https://github.com/mscdex/io.js/commit/7d0e50dcfef98ca56715adf74678bcaf4aa08796
    const result = crypto.sign(undefined, hashBuffer, key).toString('base64');
    return {
        method: 'ed25519-v2',
        public: keyPair.public,
        digest: hash,
        result,
    };
}
async function verifySignature(hash, signature) {
    (0, asserts_1.assertSignatureMethod)(signature.method);
    if (!crypto.verify) {
        throw new Error('Unable to verify signature, `crypto.verify` is not defined');
    }
    const hashBuffer = Buffer.from(hash, 'hex');
    const signatureBuffer = Buffer.from(signature.result, 'base64');
    const key = (0, format_1.importPublicKey)(signature.public);
    // The first argument must be undefined for ed25519, it defines a digest algorithm,
    // but ed25519 makes a sha512 digest as part of the algorithm. Signing and verification
    // with ed25119 doesn't work correctly if that argument is provided. Also, most crypto
    // examples use createSign(algorithm) to first create a Sign class and use that for
    // signing, but that way also doesn't work with ed25519.
    // see: https://github.com/mscdex/io.js/commit/7d0e50dcfef98ca56715adf74678bcaf4aa08796
    return crypto.verify(undefined, hashBuffer, key, signatureBuffer);
}
