// @minka/ledger-sdk/types/src/lib/records/ledger-schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaFormats = exports.SchemaFormat = exports.SchemaTargetRecords = exports.SchemaTargetRecord = void 0;
/**
 * Records that can be validated by using
 * the schema.
 */
var SchemaTargetRecord;
(function (SchemaTargetRecord) {
    SchemaTargetRecord["Anchor"] = "anchor";
    SchemaTargetRecord["AnchorLookup"] = "anchor-lookup";
    SchemaTargetRecord["Bridge"] = "bridge";
    SchemaTargetRecord["Circle"] = "circle";
    SchemaTargetRecord["Domain"] = "domain";
    SchemaTargetRecord["Effect"] = "effect";
    SchemaTargetRecord["Intent"] = "intent";
    SchemaTargetRecord["Policy"] = "policy";
    SchemaTargetRecord["Report"] = "report";
    SchemaTargetRecord["Signer"] = "signer";
    SchemaTargetRecord["SignerFactor"] = "signer-factor";
    SchemaTargetRecord["Symbol"] = "symbol";
    SchemaTargetRecord["Wallet"] = "wallet";
})(SchemaTargetRecord || (exports.SchemaTargetRecord = SchemaTargetRecord = {}));
/**
 * Schema Record values
 */
exports.SchemaTargetRecords = Object.values(SchemaTargetRecord);
/**
 * Available schema formats.
 */
var SchemaFormat;
(function (SchemaFormat) {
    SchemaFormat["JsonSchema"] = "json-schema";
})(SchemaFormat || (exports.SchemaFormat = SchemaFormat = {}));
/**
 * Schema Format values
 */
exports.SchemaFormats = Object.values(SchemaFormat);
