// @minka/ledger-sdk/types/src/lib/records/ledger-bridge.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LegacyBridgeTrait = exports.BridgeTraits = exports.BridgeTrait = exports.BridgeClaimGrouping = exports.WARNED_BRIDGE_HEADERS = exports.BLOCKED_BRIDGE_HEADERS = void 0;
/**
 * List of HTTP headers that are blocked from being set via bridge header rules
 * due to security or integrity concerns.
 *
 * This list is shared between CLI and backend to ensure consistent validation.
 */
exports.BLOCKED_BRIDGE_HEADERS = new Set([
    // Request integrity headers - can break the request or cause undefined behavior
    'host',
    'content-length',
    'transfer-encoding',
    'connection',
    'keep-alive',
    'upgrade',
    'te',
    'trailer',
    // CORS/Origin headers - manipulating these could bypass security policies
    'origin',
    'access-control-request-method',
    'access-control-request-headers',
    // Client/request metadata that should not be spoofed
    'referer',
    'user-agent',
    // Headers that could be used for IP spoofing or bypassing rate limits
    'x-forwarded-for',
    'x-forwarded-host',
    'x-forwarded-proto',
    'x-real-ip',
    'forwarded',
    // WebSocket headers - critical for handshake security
    'sec-websocket-key',
    'sec-websocket-extensions',
    'sec-websocket-accept',
    'sec-websocket-protocol',
    'sec-websocket-version',
    // HTTP/2 pseudo-headers
    ':method',
    ':path',
    ':scheme',
    ':authority',
]);
/**
 * List of headers that are potentially dangerous and should trigger a warning
 * but are not outright blocked. Users can proceed but should be careful.
 */
exports.WARNED_BRIDGE_HEADERS = new Set([
    // Can break request/response handling if set incorrectly
    'content-type',
    'content-encoding',
    'accept',
    'accept-encoding',
    'accept-language',
    // Caching headers - can cause unexpected behavior
    'cache-control',
    'pragma',
    'expires',
    'if-match',
    'if-none-match',
    'if-modified-since',
    'if-unmodified-since',
    // Range requests - can cause issues if misused
    'range',
    'if-range',
    // Expect header can cause request failures
    'expect',
]);
/**
 * Determines how bridge calls for claims are grouped in case multiple claims of
 * one intent are addressed to the same bridge.
 */
var BridgeClaimGrouping;
(function (BridgeClaimGrouping) {
    /**
     * With this option, claims will be grouped by exact address of `target.handle` of the claims.
     *
     * If there are multiple claims with the same target address,
     * they will be grouped into single credit entry sent to a target bridge.
     */
    BridgeClaimGrouping["address"] = "address";
    /**
     * With this option, claims will be grouped by the actual wallet resolved as target of the address
     * from `target.handle`
     *
     * If there are multiple claims with different `target.handle`,
     * but they get resolved to the same target wallet (e.g. because of wallet parent hierarchy),
     * these claims will be grouped together.
     *
     * Of course, if the `target.handle` is the same, then the claims will also be grouped together,
     * just like they would have with the `address` grouping setting.
     */
    BridgeClaimGrouping["wallet"] = "wallet";
})(BridgeClaimGrouping || (exports.BridgeClaimGrouping = BridgeClaimGrouping = {}));
/**
 * Features of a bridge.
 */
var BridgeTrait;
(function (BridgeTrait) {
    BridgeTrait["Debits"] = "debits";
    BridgeTrait["Credits"] = "credits";
    /**
     * Can receive intent status calls
     */
    BridgeTrait["Statuses"] = "statuses";
    BridgeTrait["Anchors"] = "anchors";
    BridgeTrait["Domains"] = "domains";
    BridgeTrait["Effects"] = "effects";
})(BridgeTrait || (exports.BridgeTrait = BridgeTrait = {}));
exports.BridgeTraits = Object.values(BridgeTrait);
/**
 * Legacy bridge features mapped to their newest versions.
 * If a bridge has a legacy trait enabled, its corresponding
 * non-legacy trait is also considered to be enabled.
 */
exports.LegacyBridgeTrait = {
    [BridgeTrait.Effects]: 'events',
};
// ^^^^^^^^^^^^^^
// TODO(Omar): We should support this for all filters.
// It basically extends LedgerFilter to support Mingo operators.
// Related to:
// https://linear.app/minka/issue/LDG-1097/fully-support-mongo-queries-in-filter-properties
