// @minka/ledger-sdk/types/src/lib/records/ledger-event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.eventSignalRecordType = exports.EventSignals = exports.EventSignal = void 0;
const dispatcher_config_1 = require("../common/dispatcher-config");
/**
 * Name of the signal of the event.
 */
var EventSignal;
(function (EventSignal) {
    EventSignal["AnchorCreated"] = "anchor-created";
    EventSignal["AnchorDropped"] = "anchor-dropped";
    EventSignal["AnchorProofsAdded"] = "anchor-proofs-added";
    EventSignal["AnchorUpdated"] = "anchor-updated";
    EventSignal["BalanceReceived"] = "balance-received";
    EventSignal["BridgeCreated"] = "bridge-created";
    EventSignal["BridgeEntryCreated"] = "bridge-entry-created";
    EventSignal["BridgeEntryProofsAdded"] = "bridge-entry-proofs-added";
    EventSignal["BridgeEntryUpdated"] = "bridge-entry-updated";
    EventSignal["BridgeProofsAdded"] = "bridge-proofs-added";
    EventSignal["BridgeUpdated"] = "bridge-updated";
    EventSignal["BridgeDropped"] = "bridge-dropped";
    EventSignal["CircleCreated"] = "circle-created";
    EventSignal["CircleProofsAdded"] = "circle-proofs-added";
    EventSignal["CircleUpdated"] = "circle-updated";
    EventSignal["DomainCreated"] = "domain-created";
    EventSignal["DomainProofsAdded"] = "domain-proofs-added";
    EventSignal["DomainUpdated"] = "domain-updated";
    EventSignal["EffectCreated"] = "effect-created";
    EventSignal["EffectProofsAdded"] = "effect-proofs-added";
    EventSignal["EffectUpdated"] = "effect-updated";
    EventSignal["EffectDropped"] = "effect-dropped";
    EventSignal["IntentCreated"] = "intent-created";
    EventSignal["IntentProofsAdded"] = "intent-proofs-added";
    EventSignal["IntentUpdated"] = "intent-updated";
    EventSignal["LedgerCreated"] = "ledger-created";
    EventSignal["LedgerDropped"] = "ledger-dropped";
    EventSignal["LedgerProofsAdded"] = "ledger-proofs-added";
    EventSignal["LedgerUpdated"] = "ledger-updated";
    EventSignal["PolicyCreated"] = "policy-created";
    EventSignal["PolicyProofsAdded"] = "policy-proofs-added";
    EventSignal["PolicyUpdated"] = "policy-updated";
    EventSignal["PolicyDropped"] = "policy-dropped";
    EventSignal["ReportCreated"] = "report-created";
    EventSignal["ReportDropped"] = "report-dropped";
    EventSignal["ReportProofsAdded"] = "report-proofs-added";
    EventSignal["ReportUpdated"] = "report-updated";
    EventSignal["RequestCreated"] = "request-created";
    EventSignal["RequestProofsAdded"] = "request-proofs-added";
    EventSignal["RequestUpdated"] = "request-updated";
    EventSignal["SchemaCreated"] = "schema-created";
    EventSignal["SchemaProofsAdded"] = "schema-proofs-added";
    EventSignal["SchemaUpdated"] = "schema-updated";
    EventSignal["SignerCreated"] = "signer-created";
    EventSignal["SignerProofsAdded"] = "signer-proofs-added";
    EventSignal["SignerUpdated"] = "signer-updated";
    EventSignal["SignerFactorCreated"] = "signer-factor-created";
    EventSignal["SignerFactorProofsAdded"] = "signer-factor-proofs-added";
    EventSignal["SignerFactorUpdated"] = "signer-factor-updated";
    EventSignal["SignerFactorDropped"] = "signer-factor-dropped";
    EventSignal["SymbolCreated"] = "symbol-created";
    EventSignal["SymbolProofsAdded"] = "symbol-proofs-added";
    EventSignal["SymbolUpdated"] = "symbol-updated";
    EventSignal["WalletCreated"] = "wallet-created";
    EventSignal["WalletDropped"] = "wallet-dropped";
    EventSignal["WalletLimited"] = "wallet-limited";
    EventSignal["WalletProofsAdded"] = "wallet-proofs-added";
    EventSignal["WalletUpdated"] = "wallet-updated";
})(EventSignal || (exports.EventSignal = EventSignal = {}));
/**
 * All possible event signals.
 */
exports.EventSignals = Object.values(EventSignal);
/**
 * Maps every event signal to the name of its primary record type.
 *
 * Using `Record<EventSignal, RecordName>` (fully exhaustive) means TypeScript
 * will produce a compile error if a new signal is added to `EventSignal`
 * without a corresponding entry here.
 */
exports.eventSignalRecordType = {
    [EventSignal.AnchorCreated]: dispatcher_config_1.RecordName.Anchor,
    [EventSignal.AnchorDropped]: dispatcher_config_1.RecordName.Anchor,
    [EventSignal.AnchorProofsAdded]: dispatcher_config_1.RecordName.Anchor,
    [EventSignal.AnchorUpdated]: dispatcher_config_1.RecordName.Anchor,
    [EventSignal.BalanceReceived]: dispatcher_config_1.RecordName.Wallet,
    [EventSignal.BridgeCreated]: dispatcher_config_1.RecordName.Bridge,
    [EventSignal.BridgeEntryCreated]: dispatcher_config_1.RecordName.BridgeEntry,
    [EventSignal.BridgeEntryProofsAdded]: dispatcher_config_1.RecordName.BridgeEntry,
    [EventSignal.BridgeEntryUpdated]: dispatcher_config_1.RecordName.BridgeEntry,
    [EventSignal.BridgeProofsAdded]: dispatcher_config_1.RecordName.Bridge,
    [EventSignal.BridgeUpdated]: dispatcher_config_1.RecordName.Bridge,
    [EventSignal.BridgeDropped]: dispatcher_config_1.RecordName.Bridge,
    [EventSignal.CircleCreated]: dispatcher_config_1.RecordName.Circle,
    [EventSignal.CircleProofsAdded]: dispatcher_config_1.RecordName.Circle,
    [EventSignal.CircleUpdated]: dispatcher_config_1.RecordName.Circle,
    [EventSignal.DomainCreated]: dispatcher_config_1.RecordName.Domain,
    [EventSignal.DomainProofsAdded]: dispatcher_config_1.RecordName.Domain,
    [EventSignal.DomainUpdated]: dispatcher_config_1.RecordName.Domain,
    [EventSignal.EffectCreated]: dispatcher_config_1.RecordName.Effect,
    [EventSignal.EffectProofsAdded]: dispatcher_config_1.RecordName.Effect,
    [EventSignal.EffectUpdated]: dispatcher_config_1.RecordName.Effect,
    [EventSignal.EffectDropped]: dispatcher_config_1.RecordName.Effect,
    [EventSignal.IntentCreated]: dispatcher_config_1.RecordName.Intent,
    [EventSignal.IntentProofsAdded]: dispatcher_config_1.RecordName.Intent,
    [EventSignal.IntentUpdated]: dispatcher_config_1.RecordName.Intent,
    [EventSignal.LedgerCreated]: dispatcher_config_1.RecordName.Ledger,
    [EventSignal.LedgerDropped]: dispatcher_config_1.RecordName.Ledger,
    [EventSignal.LedgerProofsAdded]: dispatcher_config_1.RecordName.Ledger,
    [EventSignal.LedgerUpdated]: dispatcher_config_1.RecordName.Ledger,
    [EventSignal.PolicyCreated]: dispatcher_config_1.RecordName.Policy,
    [EventSignal.PolicyProofsAdded]: dispatcher_config_1.RecordName.Policy,
    [EventSignal.PolicyUpdated]: dispatcher_config_1.RecordName.Policy,
    [EventSignal.PolicyDropped]: dispatcher_config_1.RecordName.Policy,
    [EventSignal.ReportCreated]: dispatcher_config_1.RecordName.Report,
    [EventSignal.ReportDropped]: dispatcher_config_1.RecordName.Report,
    [EventSignal.ReportProofsAdded]: dispatcher_config_1.RecordName.Report,
    [EventSignal.ReportUpdated]: dispatcher_config_1.RecordName.Report,
    [EventSignal.RequestCreated]: dispatcher_config_1.RecordName.Request,
    [EventSignal.RequestProofsAdded]: dispatcher_config_1.RecordName.Request,
    [EventSignal.RequestUpdated]: dispatcher_config_1.RecordName.Request,
    [EventSignal.SchemaCreated]: dispatcher_config_1.RecordName.Schema,
    [EventSignal.SchemaProofsAdded]: dispatcher_config_1.RecordName.Schema,
    [EventSignal.SchemaUpdated]: dispatcher_config_1.RecordName.Schema,
    [EventSignal.SignerCreated]: dispatcher_config_1.RecordName.Signer,
    [EventSignal.SignerProofsAdded]: dispatcher_config_1.RecordName.Signer,
    [EventSignal.SignerUpdated]: dispatcher_config_1.RecordName.Signer,
    [EventSignal.SignerFactorCreated]: dispatcher_config_1.RecordName.SignerFactor,
    [EventSignal.SignerFactorProofsAdded]: dispatcher_config_1.RecordName.SignerFactor,
    [EventSignal.SignerFactorUpdated]: dispatcher_config_1.RecordName.SignerFactor,
    [EventSignal.SignerFactorDropped]: dispatcher_config_1.RecordName.SignerFactor,
    [EventSignal.SymbolCreated]: dispatcher_config_1.RecordName.Symbol,
    [EventSignal.SymbolProofsAdded]: dispatcher_config_1.RecordName.Symbol,
    [EventSignal.SymbolUpdated]: dispatcher_config_1.RecordName.Symbol,
    [EventSignal.WalletCreated]: dispatcher_config_1.RecordName.Wallet,
    [EventSignal.WalletDropped]: dispatcher_config_1.RecordName.Wallet,
    [EventSignal.WalletLimited]: dispatcher_config_1.RecordName.Wallet,
    [EventSignal.WalletProofsAdded]: dispatcher_config_1.RecordName.Wallet,
    [EventSignal.WalletUpdated]: dispatcher_config_1.RecordName.Wallet,
};
