// @minka/ledger-sdk/crypto/src/lib/encryption.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decryptAES = decryptAES;
exports.encryptAES = encryptAES;
exports.exportAESEncrypted = exportAESEncrypted;
exports.importAESEncrypted = importAESEncrypted;
exports.isAESEncryptedPackage = isAESEncryptedPackage;
exports.privateDecryptRsa = privateDecryptRsa;
exports.publicEncryptRsa = publicEncryptRsa;
const detect_node_1 = __importDefault(require("detect-node"));
const lodash_1 = __importDefault(require("lodash"));
const encryption_1 = require("./node/encryption");
/**
 * Throws if the current runtime is not Node.js.
 * @param method Name of the method being called
 */
function assertNode(method) {
    if (!detect_node_1.default) {
        throw new Error(`${method} is only available in Node.js`);
    }
}
/**
 * Throws if the payload is empty, null or undefined.
 * @param method Name of the method being called
 * @param payload Payload to validate
 */
function assertPayload(method, payload) {
    if (lodash_1.default.isNil(payload)) {
        throw new Error(`Invalid value for '${method}'`);
    }
}
/**
 * Encrypts a payload using the provided RSA public key.
 * @param payload String to encrypt
 * @param publicKey RSA public key
 * @returns Base64 encoded string representing encrypted payload
 */
async function publicEncryptRsa(payload, publicKey) {
    assertNode('publicEncryptRsa');
    assertPayload('publicEncryptRsa', payload);
    return (0, encryption_1.publicEncryptRsa)(payload, publicKey);
}
/**
 * Decrypts a payload using the provided RSA private key. Returns the decrypted payload as a UTF-8 string.
 * @param encryptedPayload Base64 encoded string representing encrypted payload
 * @param privateKey Private RSA key
 * @returns UTF-8 encoded string representing decrypted payload
 */
async function privateDecryptRsa(encryptedPayload, privateKey) {
    assertNode('privateDecryptRsa');
    assertPayload('privateDecryptRsa', encryptedPayload);
    if (lodash_1.default.isEmpty(encryptedPayload)) {
        throw new Error(`Invalid value for 'privateDecryptRsa'`);
    }
    return (0, encryption_1.privateDecryptRsa)(encryptedPayload, privateKey);
}
/**
 * Encrypts a payload using the provided AES key and IV.
 * @param payload String to encrypt
 * @param aesKey Base64 encoded string or a Buffer, representing aes key
 * @param keyVersion Optional. Version of the key used for encryption. This is useful for key rotation scenarios
 * and is not handled internally by the encryption function.
 * @param ivHex Optional. Hex encoded string representing IV. If not provided, a random IV will be generated.
 * It is recommended to always use a new IV for each encryption.
 * @returns Object containing the encrypted data, IV, and authentication tag.
 */
function encryptAES(payload, aesKey, keyVersion, ivHex = undefined) {
    assertNode('encryptAES');
    assertPayload('encryptAES', payload);
    return (0, encryption_1.encryptAES)(payload, aesKey, keyVersion, ivHex);
}
/**
 * Decrypts an AES encrypted payload using the provided key and IV.
 * @param encryptedPayload Encoded string representing encrypted payload
 * @param aesKey Base64 encoded string or a Buffer, representing aes key
 * @param ivHex Hex encoded string representing IV
 * @param authTag Hex encoded string representing authentication tag
 * @returns UTF-8 encoded string representing decrypted payload
 */
function decryptAES(encryptedPayload, aesKey, ivHex, authTag) {
    assertNode('decryptAES');
    assertPayload('decryptAES', encryptedPayload);
    return (0, encryption_1.decryptAES)(encryptedPayload, aesKey, ivHex, authTag);
}
/**
 * Exports encrypted AES string components to a json string.
 * @param encryptedUnpacked Object containing the encrypted data, IV, and authentication tag.
 * @returns Encrypted string components encoded into a string.
 */
function exportAESEncrypted(encryptedUnpacked) {
    assertNode('exportAESEncrypted');
    assertPayload('exportAESEncrypted', encryptedUnpacked);
    return (0, encryption_1.exportAESEncrypted)(encryptedUnpacked);
}
/**
 * Imports encrypted string components from a string.
 * @param encryptedPacked String representing encrypted AES payload
 * @returns Encrypted string components.
 */
function importAESEncrypted(encryptedPacked) {
    assertNode('importAESEncrypted');
    assertPayload('importAESEncrypted', encryptedPacked);
    return (0, encryption_1.importAESEncrypted)(encryptedPacked);
}
/**
 * Asserts that the given string is a valid encoded `EncryptedAESPayload`
 * @param value Packed string to check
 * @returns boolean
 */
function isAESEncryptedPackage(value) {
    assertNode('isAESEncryptedPackage');
    assertPayload('isAESEncryptedPackage', value);
    return (0, encryption_1.isAESEncryptedPackage)(value);
}
