// @minka/ledger-sdk/crypto/src/lib/jwt.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decodeJWT = decodeJWT;
exports.signJWT = signJWT;
exports.getJWTProtectedHeader = getJWTProtectedHeader;
exports.verifyJWT = verifyJWT;
exports.signOauthToken = signOauthToken;
const detect_node_1 = __importDefault(require("detect-node"));
const jose_1 = require("jose");
const lodash_1 = __importDefault(require("lodash"));
const format_1 = require("../../format/src");
const jwt_1 = require("./browser/jwt");
const jwt_2 = require("./node/jwt");
/**
 * Asserts Jwt payload is compatible with the ledger Jwt
 * standard payload.
 * @param payload jwt standard payload
 * @returns parsed Ledger JWT Payload
 */
function assertLedgerJwtPayload(payload) {
    const { iss, sub, aud, jti, exp, iat, hsh, cid } = payload;
    const errors = [];
    if (!iss || !lodash_1.default.isString(iss)) {
        errors.push('"iss" must be a string');
    }
    if (!aud || !lodash_1.default.isString(aud)) {
        errors.push('"aud" must be a string');
    }
    if (jti && !lodash_1.default.isString(jti)) {
        errors.push('"jti" must be a string');
    }
    if (cid && !lodash_1.default.isString(cid)) {
        errors.push('"cid" must be a string');
    }
    if (hsh && !lodash_1.default.isString(hsh)) {
        errors.push('"hsh" must be a string');
    }
    if (!lodash_1.default.isNumber(iat) || iat <= 0) {
        errors.push('"iat" must be a positive number');
    }
    if (!lodash_1.default.isNumber(exp) || exp <= 0) {
        errors.push('"exp" must be a positive number');
    }
    if (exp < iat) {
        errors.push('"exp" must be greater than "iat"');
    }
    if (errors.length) {
        throw Error(`JWT Payload validation ${(0, format_1.pluralize)(errors.length, 'error')}: ${errors.join(',')}.`);
    }
    return {
        ...payload,
        iss,
        sub,
        aud,
        jti,
        exp,
        iat,
        hsh,
        cid,
    };
}
/**
 * Asserts Jwt protected header is compatible with the ledger Jwt
 * standard header.
 * @param header record with headers
 * @returns parsed Ledger JWT Header
 */
function assertLedgerJwtHeader(header) {
    const { kid, alg } = header;
    const errors = [];
    if (!kid || !lodash_1.default.isString(kid)) {
        errors.push('"kid" must be a string');
    }
    if (!alg || !lodash_1.default.isString(alg)) {
        errors.push('"aud" must be a string');
    }
    if (errors.length) {
        throw Error(`JWT header validation ${(0, format_1.pluralize)(errors.length, 'error')}: ${errors.join(',')}.`);
    }
    return { kid, alg };
}
/**
 * Decodes the provided JWT and returns its payload.
 * @param jwt - The signed JWT
 * @returns Ledger JWT Payload
 */
function decodeJWT(jwt) {
    /**
     * jose.decodeJwt(<jwt>) returns a stantard JWT Claims Set
     * with Registered Claim Names and a custom set of claims.
     * On the other hand, LedgerJWTPayload extends this pattern by opining
     * which of those Registered Claim Names are required
     * and naming explicitly custom claims.
     */
    const jwtPayload = (0, jose_1.decodeJwt)(jwt);
    return assertLedgerJwtPayload(jwtPayload);
}
/**
 * Sign the JWT token.
 * @param payload The JWT payload
 * @param secret The JWT secret
 * @param kid The key id
 * @returns A signed JWT token
 */
async function signJWT(payload, secret, kid) {
    return detect_node_1.default
        ? (0, jwt_2.signJWT)(payload, secret, kid)
        : (0, jwt_1.signJWT)(payload, secret, kid);
}
/**
 * Get protected header of JWT token.
 * @param jwt - The compact JWS string (header.payload.signature).
 * @returns Parsed and validated Ledger JWT header containing `kid` and `alg`.
 */
function getJWTProtectedHeader(jwt) {
    const protectedHeaders = detect_node_1.default
        ? (0, jwt_2.getJWTProtectedHeader)(jwt)
        : (0, jwt_1.getJWTProtectedHeader)(jwt);
    return assertLedgerJwtHeader(protectedHeaders);
}
/**
 * Verifies a compact JWS and returns the decoded payload and header.
 * Supports EdDSA (Ed25519) and RS256 algorithms.
 * @param jwt - The compact JWS string (header.payload.signature) to verify.
 * @param publicKey - The base64-encoded public key used to verify the signature.
 * @param format - The key format identifier. Defaults to 'ed25519-raw'.
 * @returns The decoded JWT payload and protected header.
 */
function verifyJWT(jwt, publicKey, format) {
    return detect_node_1.default
        ? (0, jwt_2.verifyJWT)(jwt, publicKey, format)
        : (0, jwt_1.verifyJWT)(jwt, publicKey, format);
}
/**
 * Signs an OAuth2 JWT using RS256 (RSA + SHA-256).
 * Routes to the Node.js or browser implementation based on runtime environment.
 * @param payload - JWT claims (iss, sub, aud, schema, iat, exp, etc.)
 * @param privateKeyPem - RSA private key in PEM format (PKCS#8)
 * @param kid - Key ID; identifies which public key verifiers should use
 * @returns A signed compact JWS string (header.payload.signature).
 */
async function signOauthToken(payload, privateKeyPem, kid) {
    return detect_node_1.default
        ? (0, jwt_2.signOauthToken)(payload, privateKeyPem, kid)
        : (0, jwt_1.signOauthToken)(payload, privateKeyPem, kid);
}
