// @minka/ledger-sdk/src/lib/policy/policy-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PolicyClient = exports.PoliciesResponse = exports.PolicyResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const drop_record_builder_1 = require("../common/services/drop-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const POLICIES_PATH = '/policies';
/**
 * Single policy API response.
 */
class PolicyResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    policy;
    constructor(response) {
        super(response);
        this.policy = response?.data?.data;
    }
}
exports.PolicyResponse = PolicyResponse;
/**
 * Schemas list API response.
 */
class PoliciesResponse extends list_response_1.BaseListResponse {
    policies;
    constructor(response) {
        super(response);
        this.policies = response?.data?.data?.map((record) => record.data);
    }
}
exports.PoliciesResponse = PoliciesResponse;
/**
 * Allows querying and managing policies.
 * @see {LedgerPolicy}
 */
class PolicyClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, POLICIES_PATH);
        this.access = new access_client_1.AccessClient(options, POLICIES_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     * @see {LedgerPolicy}
     * @see {CreateRecordBuilder}
     * @param record initial record payload (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(POLICIES_PATH, PolicyResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You need to pass the existing record payload
     * to this method to use as initial content.
     * @see {LedgerPolicy}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${POLICIES_PATH}/${encodeURIComponent(record.data.handle)}`, PolicyResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${POLICIES_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, PolicyResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets a ledger policy by identifier (luid or handle).
     * @see {LedgerPolicy}
     * @param options
     * @throws {Error} if the policy doesn't exist or the
     *   current user doesn't have access to this policy
     * @param id a unique policy handle
     * @returns a single policy response with the requested policy
     */
    async read(id, options) {
        return this.getRecord(`${POLICIES_PATH}/${encodeURIComponent(id)}`, PolicyResponse, options);
    }
    /**
     * Queries policies from a ledger.
     * @see {LedgerPolicy}
     * @param params query parameters
     * @returns policies list response with all found policies
     */
    async list(params) {
        return this.getRecordsList(POLICIES_PATH, PoliciesResponse, params);
    }
    drop(input) {
        const policyIdentifier = this.resolveIdentifier(input);
        const record = lodash_1.default.isObject(input) ? lodash_1.default.cloneDeep(input) : undefined;
        return new drop_record_builder_1.DropRecordBuilder(async (record, reqOptions) => {
            return this.dropRecord(`${POLICIES_PATH}/${encodeURIComponent(policyIdentifier)}`, record, reqOptions);
        }, async (options) => {
            const response = await this.read(policyIdentifier, options);
            return {
                data: response.policy,
                hash: response.hash,
                luid: response.luid,
                meta: response.meta,
            };
        }).init(record);
    }
}
exports.PolicyClient = PolicyClient;
