// @minka/ledger-sdk/src/lib/ledger/ledger-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LedgerClient = exports.LedgersResponse = exports.LedgerResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const drop_record_builder_1 = require("../common/services/drop-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const ledger_access_client_1 = require("./ledger-access-client");
const ledger_change_client_1 = require("./ledger-change-client");
const LEDGERS_PATH = '/ledgers';
const LEDGER_PATH = '/ledger';
/**
 * Single instance API response.
 */
class LedgerResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    ledger;
    constructor(response) {
        super(response);
        this.ledger = response?.data?.data;
    }
}
exports.LedgerResponse = LedgerResponse;
/**
 * Instances list API response.
 */
class LedgersResponse extends list_response_1.BaseListResponse {
    ledgers;
    constructor(response) {
        super(response);
        this.ledgers = response?.data?.data?.map((record) => record.data);
    }
}
exports.LedgersResponse = LedgersResponse;
/**
 * Allows querying and managing ledger instances.
 * @see {Ledger}
 */
class LedgerClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new ledger_change_client_1.LedgerChangeClient(options, LEDGER_PATH);
        this.access = new ledger_access_client_1.LedgerAccessClient(options, LEDGER_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     * @see {Ledger}
     * @see {CreateRecordBuilder}
     * @param record
     * @param options.record initial record payload (optional)
     * @param options.handle specify handle to be updated, if any (optional)
     * @returns a builder instance
     */
    init(record) {
        this.assertSingleTenant();
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(LEDGERS_PATH, LedgerResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     * @see {Ledger}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(LEDGER_PATH, LedgerResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${LEDGER_PATH}/proofs`, LedgerResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets the ledger instance from active ledger identifier
     * set in context.
     * @param options
     * @see {Ledger}
     * @throws {Error} if the instance doesn't exist or the
     *   current user doesn't have access to this instance
     * @returns a single instance response with the requested instance
     */
    async read(options) {
        return this.getRecord(LEDGER_PATH, LedgerResponse, options);
    }
    /**
     * Queries instances from a ledger.
     * @see {Ledger}
     * @param params query parameters
     * @returns instances list response with all found instances
     */
    async list(params) {
        this.assertSingleTenant();
        return this.getRecordsList(LEDGERS_PATH, LedgersResponse, params);
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     * @see {Ledger}
     * @see {DropRecordBuilder}
     * @returns a builder instance
     */
    drop() {
        const builder = new drop_record_builder_1.DropRecordBuilder(async (record, reqOptions) => {
            return this.dropRecord(LEDGER_PATH, record, reqOptions);
        }, async (options) => {
            const response = await this.read(options);
            return {
                data: response.ledger,
                hash: response.hash,
                luid: response.luid,
                meta: response.meta,
            };
        });
        return builder.init();
    }
}
exports.LedgerClient = LedgerClient;
