// @minka/ledger-sdk/src/lib/effect/activate-effect-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivateEffectBuilder = void 0;
const base_composite_builder_1 = require("../common/services/base-composite-builder");
const ledger_sdk_error_1 = require("../common/types/ledger-sdk-error");
/**
 * Specific builder for activating an effect.
 *
 * @example
 * // Activating an existing effect
 * const { effect } = await sdk.effect
 *  .with(<effect handle>)
 *  .activate({
 *    maxAge: <maxAge>
 *  })
 *  .hash()
 *  .sign([{ keyPair }])
 *  .send()
 */
class ActivateEffectBuilder extends base_composite_builder_1.BaseCompositeBuilder {
    /**
     * A handler function used to submit a signature to ledger.
     */
    sendActivation;
    /**
     * Identifier of effect on remote server i.e. handle or luid
     */
    effect;
    constructor(sendActivation, effect, payload) {
        super({ data: payload });
        this.sendActivation = sendActivation;
        this.effect = effect;
    }
    /**
     * Sends activation request to ledger.
     *
     * @param requestOptions Optional request options to use for this request
     * @returns api response
     * @throws if any method in builder chain fails
     */
    async send(requestOptions) {
        try {
            const record = await this.recordBuilder.read();
            const finalOptions = {
                authParams: requestOptions?.authParams ?? this.authParams,
                headers: requestOptions?.headers,
            };
            return this.sendActivation(record, this.effect, finalOptions);
        }
        catch (error) {
            throw new ledger_sdk_error_1.LedgerSdkError(error);
        }
    }
}
exports.ActivateEffectBuilder = ActivateEffectBuilder;
