// @minka/ledger-sdk/src/lib/effect/effect-events-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EffectEventsClient = void 0;
const base_client_1 = require("../common/clients/base-client");
const event_delivery_client_1 = require("../event-delivery/event-delivery-client");
const activate_effect_builder_1 = require("./activate-effect-builder");
const effect_client_1 = require("./effect-client");
const EVENTS_PATH = 'events';
/**
 * Allows managing effect events.
 * @see {LedgerEffect}
 */
class EffectEventsClient extends base_client_1.BaseClient {
    parentPath;
    clientOptions;
    constructor(options, parentPath) {
        super(options);
        this.parentPath = parentPath;
        this.clientOptions = options;
    }
    buildBasePath() {
        const address = this.assertParentId();
        return `${this.parentPath}/${encodeURIComponent(address)}/${EVENTS_PATH}`;
    }
    /**
     * Lists event delivery records for this effect.
     * @param params optional filter and pagination parameters
     * @returns list response with event delivery records
     */
    async list(params) {
        return this.getRecordsList(this.buildBasePath(), event_delivery_client_1.EventDeliveryListResponse, params);
    }
    /**
     * Fetches a single event delivery record by its public handle.
     * 404 when the handle doesn't belong to this effect.
     * @param deliveryHandle - Public delivery handle (from `list().data.handle`).
     * @returns single event delivery record response.
     */
    async find(deliveryHandle) {
        return this.getRecord(`${this.buildBasePath()}/${encodeURIComponent(deliveryHandle)}`, event_delivery_client_1.EventDeliveryResponse);
    }
    /**
     * Initiates bulk reactivation of failed events for this effect.
     * @param payload
     * @example
     * await sdk.effect
     *   .with(effectHandle)
     *   .events.retry({ maxAge: 60 })
     *   .hash()
     *   .sign([{ keyPair }])
     *   .send()
     */
    retry(payload) {
        const id = this.assertParentId();
        return new activate_effect_builder_1.ActivateEffectBuilder(async (record, effect, reqOptions) => {
            return this.activateRecord(`${this.buildBasePath()}/retry`, effect_client_1.EffectActivationResponse, record, reqOptions);
        }, id, payload);
    }
}
exports.EffectEventsClient = EffectEventsClient;
