// @minka/ledger-sdk/src/lib/common/services/check-access.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessCheckBuilder = void 0;
const ledger_sdk_error_1 = require("../types/ledger-sdk-error");
const base_record_builder_1 = require("./base-record-builder");
/**
 * Implements a builder pattern that allows a more
 * convenient way to work with ledger records. This
 * class has functions that simplify construction,
 * hashing, signing, testing and sending records to
 * remote ledger instances.
 *
 * This builder is intended for sending access checks
 * to ledger, and will attempt to send it on the API
 * with the content that was prepared in the builder.
 *
 * NOTE: current access check is mutated by this
 * class, we are not cloning instances intentionally,
 * and it isn't advised to reuse builder instances when
 * working with multiple records or mutating objects
 * passed to this builder in external code. It is
 * recommended to clone objects in such cases to avoid
 * unexpected behaviors.
 */
class AccessCheckBuilder extends base_record_builder_1.BaseRecordBuilder {
    /**
     * A handler for access check on the remote API.
     */
    checkAccess;
    /**
     * Creates new instance of access check builder.
     *
     * @param checkAccess function which handles access check on ledger api
     */
    constructor(checkAccess) {
        super();
        this.checkAccess = checkAccess;
    }
    /**
     * Sends the current access check payload to the remote ledger.
     *
     * @param requestOptions Optional request options to use for this request
     * @returns api response with the checked access response
     * @throws if any method in builder chain fails
     */
    async send(requestOptions) {
        try {
            await this.finishAsyncOperations();
            const finalOptions = {
                authParams: requestOptions?.authParams ?? this.authParams,
                headers: requestOptions?.headers,
            };
            return this.checkAccess(this.record, finalOptions);
        }
        catch (error) {
            throw new ledger_sdk_error_1.LedgerSdkError(error);
        }
    }
}
exports.AccessCheckBuilder = AccessCheckBuilder;
