// @minka/ledger-sdk/src/lib/common/services/drop-record-builder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropRecordBuilder = void 0;
const lodash_1 = __importDefault(require("lodash"));
const types_1 = require("../../../../types/src");
const ledger_sdk_error_1 = require("../types/ledger-sdk-error");
const base_record_builder_1 = require("./base-record-builder");
/**
 * Implements a builder pattern that allows a more
 * convenient way to work with ledger records. This
 * class has functions that simplify construction,
 * hashing, signing, testing and sending records to
 * remote ledger instances.
 *
 * This builder is intended for record drops, sending
 * a record to ledger will perform the drop operation
 * on the API with the record that was prepared in the
 * builder.
 *
 * NOTE: current record instance is mutated by this
 * class, we are not cloning instances intentionally,
 * and it isn't advised to reuse builder instances when
 * working with multiple records or mutating objects
 * passed to this builder in external code. It is
 * recommended to clone objects in such cases to avoid
 * unexpected behaviors.
 */
class DropRecordBuilder extends base_record_builder_1.BaseRecordBuilder {
    /**
     * A handler for droping records on the remote API.
     */
    dropRecord;
    /**
     * A handler for reading records on the remote API.
     */
    readRecord;
    /**
     * Creates new instance of record builder.
     * @param dropRecord function which drops records on ledger API.
     * @param readRecord function which reads records from ledger API.
     * @param existingRecord record to initialize the builder with (optional) to avoid an extra read call.
     * @param record
     */
    constructor(dropRecord, readRecord) {
        super();
        this.dropRecord = dropRecord;
        this.readRecord = readRecord;
    }
    /**
     * Initializes the record.
     * @param record the record
     * @returns this builder instance for chaining
     */
    init(record) {
        this.enqueueAsyncOperation(async () => {
            if (!record) {
                this.record = await this.readRecord({ authParams: this.authParams });
            }
            else {
                this.record = lodash_1.default.cloneDeep(record);
            }
            lodash_1.default.set(this.record, 'meta.moment', undefined);
        });
        return this;
    }
    /**
     * Sends the current record version to the remote ledger.
     * @param requestOptions Optional request options to use for this request
     * @returns api response with the dropped record
     * @throws if any method in builder chain fails
     */
    async send(requestOptions) {
        try {
            await this.finishAsyncOperations();
            const finalOptions = {
                authParams: requestOptions?.authParams ?? this.authParams,
                headers: requestOptions?.headers,
            };
            return this.dropRecord(this.record, finalOptions);
        }
        catch (error) {
            throw new ledger_sdk_error_1.LedgerSdkError(error);
        }
    }
    /**
     * Assigns data.parent with current hash.
     * Proceeds to call parent hash() implementation to recalculate the hash.
     * @param hash the hash to set, or a falsy value to calculate a new hash
     * @returns this builder instance for chaining
     */
    hash(hash) {
        this.enqueueAsyncOperation(async () => {
            lodash_1.default.set(this.record, 'data', { parent: this.record.hash });
        });
        return super.hash(hash);
    }
    /**
     * sets custom.status in all signatures to the ones provided in custom.status of params, or falls back to status 'dropped'
     * @param params signatures
     * @returns this builder instance for chaining
     */
    sign(params) {
        this.enqueueAsyncOperation(async () => {
            this.record.meta.proofs = [];
        });
        const signingParams = lodash_1.default.map(params, (param) => {
            const updatedParam = structuredClone(param);
            if (!param?.custom?.status) {
                lodash_1.default.set(updatedParam, 'custom.status', types_1.LedgerSignatureStatus.Dropped);
            }
            return updatedParam;
        });
        return super.sign(signingParams);
    }
}
exports.DropRecordBuilder = DropRecordBuilder;
