// @minka/ledger-sdk/src/lib/anchor/anchor-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnchorClient = exports.AnchorsResponse = exports.AnchorResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const drop_record_builder_1 = require("../common/services/drop-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const ANCHORS_PATH = '/anchors';
/**
 * Single anchor API response.
 */
class AnchorResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    anchor;
    constructor(response) {
        super(response);
        this.anchor = response?.data?.data;
    }
}
exports.AnchorResponse = AnchorResponse;
/**
 * Anchors list API response.
 */
class AnchorsResponse extends list_response_1.BaseListResponse {
    /**
     * A flattened list of records, extracted from the `data` property of
     * each record returned.
     */
    anchors;
    constructor(response) {
        super(response);
        this.anchors = response?.data?.data?.map((record) => record.data);
    }
}
exports.AnchorsResponse = AnchorsResponse;
/**
 * Allows querying and managing anchors.
 * @see {LedgerAnchor}
 */
class AnchorClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, ANCHORS_PATH);
        this.access = new access_client_1.AccessClient(options, ANCHORS_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     * @see {LedgerAnchor}
     * @see {CreateRecordBuilder}
     * @param record initial record payload (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(ANCHORS_PATH, AnchorResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     * @see {LedgerAnchor}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${ANCHORS_PATH}/${encodeURIComponent(record.data.handle)}`, AnchorResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${ANCHORS_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, AnchorResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    drop(input) {
        const anchorIdentifier = this.resolveIdentifier(input);
        const record = lodash_1.default.isObject(input) ? lodash_1.default.cloneDeep(input) : undefined;
        return new drop_record_builder_1.DropRecordBuilder(async (record, reqOptions) => {
            return this.dropRecord(`${ANCHORS_PATH}/${encodeURIComponent(anchorIdentifier)}`, record, reqOptions);
        }, async (options) => {
            const response = await this.read(anchorIdentifier, options);
            return {
                data: response.anchor,
                hash: response.hash,
                luid: response.luid,
                meta: response.meta,
            };
        }).init(record);
    }
    /**
     * Gets a ledger anchor by identifier (luid or handle).
     * @see {LedgerAnchor}
     * @param id a unique anchor handle
     * @param options read options (optional)
     * @throws {Error} if the anchor doesn't exist or the
     *   current user doesn't have access to this anchor
     * @returns a single anchor response with the requested anchor
     */
    async read(id, options) {
        return this.getRecord(`${ANCHORS_PATH}/${encodeURIComponent(id)}`, AnchorResponse, options);
    }
    /**
     * Queries anchors from a ledger.
     * @see {LedgerAnchor}
     * @param params query parameters (optional)
     * @returns anchors list response with all found anchors
     */
    async list(params) {
        return this.getRecordsList(ANCHORS_PATH, AnchorsResponse, params);
    }
}
exports.AnchorClient = AnchorClient;
