// @minka/ledger-sdk/src/lib/schema/schema-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaClient = exports.SchemasResponse = exports.SchemaResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const list_response_1 = require("../common/types/list-response");
const SCHEMAS_PATH = '/schemas';
/**
 * Single schema API response.
 */
class SchemaResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    schema;
    constructor(response) {
        super(response);
        this.schema = response?.data?.data;
    }
}
exports.SchemaResponse = SchemaResponse;
/**
 * Schemas list API response.
 */
class SchemasResponse extends list_response_1.BaseListResponse {
    schemas;
    constructor(response) {
        super(response);
        this.schemas = response?.data?.data?.map((record) => record.data);
    }
}
exports.SchemasResponse = SchemasResponse;
/**
 * Allows querying and managing schemas.
 *
 * @see {LedgerSchema}
 */
class SchemaClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, SCHEMAS_PATH);
        this.access = new access_client_1.AccessClient(options, SCHEMAS_PATH);
        this.links = [this.change, this.access];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     *
     * @see {LedgerSchema}
     * @see {CreateRecordBuilder}
     * @param record initial record payload (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(SCHEMAS_PATH, SchemaResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You need to pass the existing record payload
     * to this method to use as initial content.
     *
     * @see {LedgerSchema}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${SCHEMAS_PATH}/${encodeURIComponent(record.data.handle)}`, SchemaResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${SCHEMAS_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, SchemaResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets a ledger schema by identifier (luid or handle).
     *
     * @see {LedgerSchema}
     * @throws {Error} if the schema doesn't exist or the
     *   current user doesn't have access to this schema
     * @param id a unique schema handle
     * @returns a single schema response with the requested schema
     */
    async read(id, options) {
        return this.getRecord(`${SCHEMAS_PATH}/${encodeURIComponent(id)}`, SchemaResponse, options);
    }
    /**
     * Queries schemas from a ledger.
     *
     * @see {LedgerSchema}
     * @param params query parameters
     * @returns schemas list response with all found schemas
     */
    async list(params) {
        return this.getRecordsList(SCHEMAS_PATH, SchemasResponse, params);
    }
}
exports.SchemaClient = SchemaClient;
