// @minka/ledger-sdk/src/lib/oauth/oauth-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuthClient = void 0;
const base_client_1 = require("../common/clients/base-client");
/**
 * Allows exchanging OAuth 2.0 client credentials for an access token.
 *
 * This client calls the non-standard `/oauth/token` endpoint which returns
 * `{ access_token, token_type, expires_in }` rather than the ledger's
 * standard `{ hash, data, meta }` format. It therefore calls the underlying
 * axios instance directly instead of using the base record methods.
 */
class OAuthClient extends base_client_1.BaseClient {
    /**
     * Exchanges client credentials for an OAuth 2.0 access token.
     *
     * Implements RFC 6749 Section 4.4 (Client Credentials Grant).
     * The returned `access_token` can be passed to `sdk.setAuthParams({ overrideToken })`
     * to authenticate all subsequent SDK calls without managing cryptographic keys.
     * @param clientId - Client ID from the signer's `oauth-client-credentials` factor
     * @param clientSecret - Client secret revealed once at factor creation via `include: ['meta.secret']`
     * @returns `OAuthTokenResponse` with `access_token`, `token_type`, and `expires_in`
     */
    async exchangeToken(clientId, clientSecret) {
        const ledger = this.getActiveLedger();
        const credentials = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
        const { data } = await this.apiClient.post('/oauth/token', new URLSearchParams({ grant_type: 'client_credentials' }), {
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
                Authorization: `Basic ${credentials}`,
                ...(ledger ? { 'X-Ledger': ledger } : {}),
            },
        });
        return {
            accessToken: data.access_token,
            tokenType: data.token_type,
            expiresIn: data.expires_in,
        };
    }
    /**
     * No-op — this client has no ledger record to initialize.
     */
    init() {
        return null;
    }
}
exports.OAuthClient = OAuthClient;
