// @minka/ledger-sdk/src/lib/bridge/bridge-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeClient = exports.BridgesResponse = exports.BridgeActivationResponse = exports.BridgeResponse = void 0;
const lodash_1 = __importDefault(require("lodash"));
const access_client_1 = require("../common/clients/access-client");
const base_change_client_1 = require("../common/clients/base-change-client");
const base_client_1 = require("../common/clients/base-client");
const create_record_builder_1 = require("../common/services/create-record-builder");
const drop_record_builder_1 = require("../common/services/drop-record-builder");
const update_record_builder_1 = require("../common/services/update-record-builder");
const base_record_response_1 = require("../common/types/base-record-response");
const base_response_1 = require("../common/types/base-response");
const list_response_1 = require("../common/types/list-response");
const activate_bridge_builder_1 = require("./activate-bridge-builder");
const bridge_events_client_1 = require("./bridge-events-client");
const BRIDGES_PATH = '/bridges';
/**
 * Single bridge API response.
 */
class BridgeResponse extends base_record_response_1.BaseRecordResponse {
    /**
     * Record data extracted from the `data` property of the response body.
     */
    bridge;
    constructor(response) {
        super(response);
        this.bridge = response?.data?.data;
    }
}
exports.BridgeResponse = BridgeResponse;
/**
 * Bridge activation API response.
 */
// TODO: return a signed response from ledger with hash and meta
class BridgeActivationResponse extends base_response_1.BaseResponse {
}
exports.BridgeActivationResponse = BridgeActivationResponse;
/**
 * Bridges list API response.
 */
class BridgesResponse extends list_response_1.BaseListResponse {
    /**
     * A flattened list of records, extracted from the `data` property of
     * each record returned.
     */
    bridges;
    constructor(response) {
        super(response);
        this.bridges = response?.data?.data?.map((record) => record.data);
    }
}
exports.BridgesResponse = BridgesResponse;
/**
 * Allows querying and managing bridges.
 * @see {LedgerBridge}
 */
class BridgeClient extends base_client_1.BaseClient {
    /**
     * Allows querying changes.
     *
     */
    change;
    /**
     * Allows checking access.
     *
     */
    access;
    /**
     * Allows managing events.
     *
     */
    events;
    constructor(options) {
        super(options);
        this.change = new base_change_client_1.BaseChangeClient(options, BRIDGES_PATH);
        this.access = new access_client_1.AccessClient(options, BRIDGES_PATH);
        this.events = new bridge_events_client_1.BridgeEventsClient(options, BRIDGES_PATH);
        this.links = [this.change, this.access, this.events];
    }
    /**
     * Initializes a builder instance that supports chaining to make it
     * easier to work with ledger records. You can pass the record
     * payload to this method to use as initial content, this is useful
     * when you already have an object ready, for example when validating
     * API responses or modifying an existing record. An empty record is
     * going to be created if no initial data is provided.
     * @see {LedgerBridge}
     * @see {CreateRecordBuilder}
     * @param record initial record payload (optional)
     * @returns a builder instance
     */
    init(record) {
        const builder = new create_record_builder_1.CreateRecordBuilder(async (record, reqOptions) => {
            return this.createRecord(BRIDGES_PATH, BridgeResponse, record, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Initializes a builder instance for activating existing record.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the new record payload to this
     * method to use as initial content.
     * @param payload initial activation payload
     * @returns a builder instance
     */
    activate(payload) {
        const id = this.assertId();
        return new activate_bridge_builder_1.ActivateBridgeBuilder(async (record, bridge, reqOptions) => {
            return this.activateRecord(`${BRIDGES_PATH}/${encodeURIComponent(bridge)}/activate`, BridgeActivationResponse, record, reqOptions);
        }, id, lodash_1.default.cloneDeep(payload));
    }
    /**
     * Initializes a builder instance for updating existing records.
     * The builder supports chaining to make it easier to work with
     * ledger records. You can pass the existing record payload to
     * this method to use as initial content.
     * @see {LedgerBridge}
     * @see {UpdateRecordBuilder}
     * @param record initial record payload
     * @returns a builder instance
     */
    from(record) {
        const builder = new update_record_builder_1.UpdateRecordBuilder(async (record, reqOptions) => {
            return this.updateRecord(`${BRIDGES_PATH}/${encodeURIComponent(record.data.handle)}`, BridgeResponse, record, reqOptions);
        }, async (proof, reqOptions) => {
            return this.signRecord(`${BRIDGES_PATH}/${encodeURIComponent(record.data.handle)}/proofs`, BridgeResponse, proof, reqOptions);
        });
        record = lodash_1.default.cloneDeep(record);
        return builder.init(record);
    }
    /**
     * Gets a ledger bridge by identifier (luid or handle).
     * @see {LedgerBridge}
     * @param id a unique bridge handle
     * @param options read options (optional)
     * @throws {Error} if the bridge doesn't exist or the
     *   current user doesn't have access to this bridge
     * @returns a single bridge response with the requested bridge
     */
    async read(id, options) {
        return this.getRecord(`${BRIDGES_PATH}/${encodeURIComponent(id)}`, BridgeResponse, options);
    }
    /**
     * Queries bridges from a ledger.
     * @see {LedgerBridge}
     * @param params query parameters
     * @returns bridges list response with all found bridges
     */
    async list(params) {
        return this.getRecordsList(BRIDGES_PATH, BridgesResponse, params);
    }
    drop(input) {
        const bridgeIdentifier = this.resolveIdentifier(input);
        const record = lodash_1.default.isObject(input) ? lodash_1.default.cloneDeep(input) : undefined;
        return new drop_record_builder_1.DropRecordBuilder(async (record, reqOptions) => {
            return this.dropRecord(`${BRIDGES_PATH}/${encodeURIComponent(bridgeIdentifier)}`, record, reqOptions);
        }, async (options) => {
            const response = await this.read(bridgeIdentifier, options);
            return {
                data: response.bridge,
                hash: response.hash,
                luid: response.luid,
                meta: response.meta,
            };
        }).init(record);
    }
}
exports.BridgeClient = BridgeClient;
