// @minka/ledger-sdk/src/lib/bridge/bridge-events-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BridgeEventsClient = void 0;
const base_client_1 = require("../common/clients/base-client");
const event_delivery_client_1 = require("../event-delivery/event-delivery-client");
const activate_bridge_builder_1 = require("./activate-bridge-builder");
const bridge_client_1 = require("./bridge-client");
const EVENTS_PATH = 'events';
/**
 * Allows managing bridge events.
 * @see {LedgerBridge}
 */
class BridgeEventsClient extends base_client_1.BaseClient {
    parentPath;
    clientOptions;
    constructor(options, parentPath) {
        super(options);
        this.parentPath = parentPath;
        this.clientOptions = options;
    }
    buildBasePath() {
        const address = this.assertParentId();
        return `${this.parentPath}/${encodeURIComponent(address)}/${EVENTS_PATH}`;
    }
    /**
     * Lists event delivery records for this bridge.
     * @param params optional filter and pagination parameters
     * @returns list response with event delivery records
     */
    async list(params) {
        return this.getRecordsList(this.buildBasePath(), event_delivery_client_1.EventDeliveryListResponse, params);
    }
    /**
     * Fetches a single event delivery record by its public handle.
     * 404 when the handle doesn't belong to this bridge.
     * @param deliveryHandle - Public delivery handle (from `list().data.handle`).
     * @returns single event delivery record response.
     */
    async find(deliveryHandle) {
        return this.getRecord(`${this.buildBasePath()}/${encodeURIComponent(deliveryHandle)}`, event_delivery_client_1.EventDeliveryResponse);
    }
    /**
     * Initiates bulk reactivation of failed events for this bridge.
     * @param payload
     * @example
     * await sdk.bridge
     *   .with(bridgeHandle)
     *   .events.retry({ maxAge: 60 })
     *   .hash()
     *   .sign([{ keyPair }])
     *   .send()
     */
    retry(payload) {
        const id = this.assertParentId();
        return new activate_bridge_builder_1.ActivateBridgeBuilder(async (record, bridge, reqOptions) => {
            return this.activateRecord(`${this.buildBasePath()}/retry`, bridge_client_1.BridgeActivationResponse, record, reqOptions);
        }, id, payload);
    }
}
exports.BridgeEventsClient = BridgeEventsClient;
